theano/gof/opt.py

```diff
@@ -1353,21 +1353,27 @@ def transform(self, node):
                     self.process_count[opt] += 1
                 if not new_repl:
                     continue
-                else:
-                    if self.profile:
-                        self.node_created[opt] += len(graph.ops(fgraph.variables, new_repl))
-                        self.applied_true[opt] += 1
-                    break  # break from the for loop over optimization.
+                if isinstance(new_repl, (tuple, list)):
+                    new_vars = new_repl
+                else:  # It must be a dict
+                    new_vars = list(new_repl.values())
```

Comment:
I just learned that optimizer have support for handling dict replacements from LocalOptimizer by reading the code.  I don't believe I ever saw a single optimization in Theano use that.  So this won't replace random things.

However, with the non-dict code we always try again with the owner of the first output.  With the dict code, it may be one or the other, depending in which order they end up in the list.

Also the commit messages made it sound like this was necessary to remove nodes from the graph, but it's not since optimizers can include `None` in the list to remove certain outputs.